File: src/backend/ollama-installed.ts
import { createHash, randomUUID } from "node:crypto";
import { createReadStream } from "node:fs";
import { readFile, stat } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";
import { z } from "zod";
import { BackendError, ValidationError } from "../errors.js";
import { assertLoopbackEndpoint } from "./adapter.js";
import { assertSafeModelId } from "./net.js";
import type { InstalledModel, InstalledModelSupport } from "./installed.js";

const DIGEST = z.string().regex(/^[a-f0-9]{64}$/);
const TAGS = z.object({
  models: z
    .array(
      z.object({
        name: z.string().min(1).max(256),
        digest: DIGEST,
        size: z.number().int().positive().safe(),
        remote_host: z.string().optional(),
        remote_model: z.string().optional(),
        details: z.object({ quantization_level: z.string().max(100).optional() }).optional(),
      }),
    )
    .max(10000),
});
const SHOW = z.object({
  model_info: z.record(z.unknown()).default({}),
  capabilities: z.array(z.string().max(100)).max(100).default([]),
  parameters: z.string().max(100000).optional(),
  remote_host: z.string().optional(),
});
const CONTEXT = z.number().int().min(1).max(10_000_000);
const LAYER = z.object({
  digest: z.string().regex(/^sha256:[a-f0-9]{64}$/),
  size: z.number().int().nonnegative().safe(),
  mediaType: z.string(),
});
const MANIFEST = z.object({
  schemaVersion: z.literal(2),
  config: LAYER,
  layers: z.array(LAYER).min(1).max(10000),
});

export interface InstalledVerifierOptions {
  readonly modelsDir?: string;
  readonly readFile?: (path: string) => Promise<string>;
  readonly hashFile?: (path: string) => Promise<string>;
  readonly statFile?: (path: string) => Promise<{ size: number }>;
}

async function hashBlob(path: string): Promise<string> {
  const hash = createHash("sha256");
  for await (const chunk of createReadStream(path)) hash.update(chunk as Buffer);
  return hash.digest("hex");
}

export function createInstalledModelVerifier(
  options: InstalledVerifierOptions = {},
): InstalledModelSupport["verify"] {
  const root =
    options.modelsDir ?? process.env["OLLAMA_MODELS"] ?? join(homedir(), ".ollama", "models");
  const read = options.readFile ?? ((path: string) => readFile(path, "utf8"));
  const hash = options.hashFile ?? hashBlob;
  const measure = options.statFile ?? stat;
  return async (modelId, digest, expectedSha256, expectedSizeBytes) => {
    assertSafeModelId(modelId);
    if (!DIGEST.safeParse(digest).success) throw new ValidationError("invalid manifest digest");
    const segments = modelId.split("/");
    if (segments.length > 3) throw new ValidationError("invalid installed model path");
    const last = segments.pop()!;
    const parts = last.split(":");
    const name = parts[0]!;
    const tag = parts[1] ?? "latest";
    const namespace = segments.pop() ?? "library";
    const registry = segments.pop() ?? "registry.ollama.ai";
    if (
      parts.length > 2 ||
      ![registry, namespace, name, tag].every((part) => /^[a-z0-9][a-z0-9._-]*$/.test(part))
    ) {
      throw new ValidationError("invalid installed model path");
    }
    const path = join(root, "manifests", registry, namespace, name, tag);
    try {
      const raw = await read(path);
      if (
        raw.length > 4 * 1024 * 1024 ||
        createHash("sha256").update(raw).digest("hex") !== digest
      ) {
        throw new BackendError("local manifest integrity mismatch");
      }
      const manifest = MANIFEST.parse(JSON.parse(raw) as unknown);
      const weights = manifest.layers.filter(
        (layer) => layer.mediaType === "application/vnd.ollama.image.model",
      );
      if (
        expectedSha256 !== undefined &&
        !weights.some((layer) => layer.digest === `sha256:${expectedSha256}`)
      ) {
        throw new BackendError("installed weights do not match the catalog digest");
      }
      if (
        expectedSizeBytes !== undefined &&
        weights.reduce((total, layer) => total + layer.size, 0) < expectedSizeBytes * 0.5
      ) {
        throw new BackendError("installed weights do not meet the catalog size floor");
      }
      if (
        !manifest.layers.some((layer) => layer.mediaType === "application/vnd.ollama.image.model")
      ) {
        throw new BackendError("local manifest has no model weights");
      }
      for (const layer of [manifest.config, ...manifest.layers]) {
        const blob = join(root, "blobs", layer.digest.replace(":", "-"));
        if (
          (await measure(blob)).size !== layer.size ||
          (await hash(blob)) !== layer.digest.slice(7)
        ) {
          throw new BackendError("local blob integrity mismatch");
        }
      }
      if ((await read(path)) !== raw)
        throw new BackendError("local manifest changed during verification");
    } catch (cause) {
      throw new BackendError(
        `installed model integrity verification failed for ${modelId}: ${cause instanceof Error ? cause.message : "unavailable content"}`,
        { cause },
      );
    }
  };
}

export interface OllamaInstalledOptions {
  readonly request: (
    endpoint: string,
    path: string,
    body?: Record<string, unknown>,
  ) => Promise<unknown>;
  readonly verify: InstalledModelSupport["verify"];
  readonly uniqueId?: () => string;
}

export class OllamaInstalledModels implements InstalledModelSupport {
  readonly verify: InstalledModelSupport["verify"];
  constructor(private readonly deps: OllamaInstalledOptions) {
    this.verify = deps.verify;
  }

  async list(endpoint: string): Promise<readonly InstalledModel[]> {
    endpoint = assertLoopbackEndpoint(endpoint);
    const parsed = TAGS.safeParse(await this.deps.request(endpoint, "/api/tags"));
    if (!parsed.success)
      throw new BackendError("invalid installed model inventory", { cause: parsed.error });
    return parsed.data.models
      .filter((entry) => !entry.remote_host && !entry.remote_model)
      .map((entry) => {
        assertSafeModelId(entry.name);
        return {
          id: entry.name,
          digest: entry.digest,
          sizeBytes: entry.size,
          quant: entry.details?.quantization_level ?? null,
          contextLength: null,
          kvBytesPerToken: null,
          capabilities: [],
        };
      });
  }

  async inspect(endpoint: string, modelId: string): Promise<InstalledModel> {
    assertSafeModelId(modelId);
    const selected = (await this.list(endpoint)).find((entry) => entry.id === modelId);
    if (selected === undefined)
      throw new ValidationError(`model ${modelId} is not installed locally`);
    const result = SHOW.safeParse(
      await this.deps.request(assertLoopbackEndpoint(endpoint), "/api/show", { model: modelId }),
    );
    if (!result.success || result.data.remote_host)
      throw new BackendError("invalid local model metadata");
    const info = result.data.model_info;
    const architecture = z.string().safeParse(info["general.architecture"]);
    const family = architecture.success ? architecture.data : "";
    const context = CONTEXT.safeParse(info[`${family}.context_length`]);
    let kvBytesPerToken: number | null = null;
    if (["llama", "qwen2"].includes(family)) {
      const geometry = z
        .array(z.number().int().positive().max(100000))
        .length(4)
        .safeParse([
          info[`${family}.block_count`],
          info[`${family}.attention.head_count_kv`],
          info[`${family}.attention.key_length`],
          info[`${family}.attention.value_length`],
        ]);
      if (geometry.success) {
        const [layers, heads, key, value] = geometry.data as [number, number, number, number];
        kvBytesPerToken = layers * heads * (key + value) * 2;
      }
    }
    return {
      ...selected,
      contextLength: context.success ? context.data : null,
      kvBytesPerToken,
      capabilities: result.data.capabilities,
    };
  }

  async activate(endpoint: string, model: InstalledModel, context?: number): Promise<string> {
    endpoint = assertLoopbackEndpoint(endpoint);
    if (context !== undefined && !CONTEXT.safeParse(context).success)
      throw new ValidationError("invalid context");
    await this.verify(model.id, model.digest);
    const current = (await this.list(endpoint)).find((entry) => entry.id === model.id);
    if (current?.digest !== model.digest)
      throw new BackendError("installed model manifest changed; retry");
    if (context === undefined) return model.id;
    const variant = `llmup-context-${(this.deps.uniqueId ?? randomUUID)()}:${String(context)}`;
    assertSafeModelId(variant);
    const created = z.object({ status: z.literal("success") }).safeParse(
      await this.deps.request(endpoint, "/api/create", {
        model: variant,
        from: model.id,
        parameters: { num_ctx: context },
        stream: false,
      }),
    );
    if (!created.success) throw new BackendError("context variant creation failed");
    const metadata = SHOW.safeParse(
      await this.deps.request(endpoint, "/api/show", { model: variant }),
    );
    if (
      !metadata.success ||
      !metadata.data.parameters?.split("\n").some((line) => {
        const [name, value] = line.trim().split(/\s+/);
        return name === "num_ctx" && Number(value) === context;
      })
    )
      throw new BackendError("runtime did not retain requested context");
    const variantModel = (await this.list(endpoint)).find((entry) => entry.id === variant);
    if (variantModel === undefined) throw new BackendError("context variant is not installed");
    await this.verify(variantModel.id, variantModel.digest);
    const after = (await this.list(endpoint)).find((entry) => entry.id === model.id);
    if (after?.digest !== model.digest)
      throw new BackendError("source model changed while configuring context");
    return variant;
  }
}

File: crates/llmup-runtime/src/ollama_installed.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
};
use tokio_util::sync::CancellationToken;

#[derive(Debug, thiserror::Error)]
pub enum OllamaError {
    #[error("invalid Ollama input or response: {0}")]
    Invalid(&'static str),
    #[error("installed model integrity verification failed: {0}")]
    Integrity(&'static str),
    #[error("Ollama request failed: {0}")]
    Request(String),
}
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InstalledModel {
    pub id: String,
    pub digest: String,
    pub size_bytes: u64,
    pub quant: Option<String>,
    pub context_length: Option<u32>,
    pub kv_bytes_per_token: Option<u64>,
    pub capabilities: Vec<String>,
}
pub fn size_installed(
    model: &InstalledModel,
    hardware: &llmup_core::sizing::Hardware,
    context: Option<u32>,
) -> Result<Value, OllamaError> {
    crate::hardware::validate_hardware(hardware).map_err(|_| OllamaError::Invalid("hardware"))?;
    if context.is_some_and(|value| !(1..=10000000).contains(&value)) {
        return Err(OllamaError::Invalid("context"));
    }
    let (kind, capacity) = llmup_core::sizing::memory_capacity(hardware);
    let usable = capacity * 0.85;
    let weights_fit = model.size_bytes as f64 <= usable;
    let required = context
        .zip(model.kv_bytes_per_token)
        .map(|(context, kv)| model.size_bytes as f64 + kv as f64 * f64::from(context));
    let over_cap = context
        .zip(model.context_length)
        .is_some_and(|(context, cap)| context > cap);
    let fit = if over_cap || !weights_fit || required.is_some_and(|required| required > usable) {
        "no"
    } else if required.is_none() {
        "unknown"
    } else {
        "yes"
    };
    let mut value =
        serde_json::to_value(model).map_err(|_| OllamaError::Invalid("installed model"))?;
    for (key, entry) in [
        ("context", serde_json::json!(context)),
        ("fit", serde_json::json!(fit)),
        ("weightsFit", serde_json::json!(weights_fit)),
        ("requiredBytes", serde_json::json!(required)),
        ("usableBytes", serde_json::json!(usable)),
        ("memoryKind", serde_json::json!(kind)),
        ("evidence", serde_json::json!("local-runtime-metadata")),
        ("throughput", serde_json::json!("unknown")),
    ] {
        value[key] = entry;
    }
    Ok(value)
}
fn digest(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}
pub(crate) fn model_path(root: &Path, id: &str) -> Result<PathBuf, OllamaError> {
    let mut segments: Vec<_> = id.split('/').collect();
    if segments.len() > 3 {
        return Err(OllamaError::Invalid("installed model path"));
    }
    let last = segments.pop().ok_or(OllamaError::Invalid("model id"))?;
    let parts: Vec<_> = last.split(':').collect();
    if parts.len() > 2 {
        return Err(OllamaError::Invalid("model tag"));
    }
    let name = parts[0];
    let tag = parts.get(1).copied().unwrap_or("latest");
    let namespace = segments.pop().unwrap_or("library");
    let registry = segments.pop().unwrap_or("registry.ollama.ai");
    let safe = |value: &str| {
        !value.is_empty()
            && value.as_bytes()[0].is_ascii_alphanumeric()
            && value.bytes().all(|byte| {
                byte.is_ascii_lowercase() || byte.is_ascii_digit() || b"._-".contains(&byte)
            })
    };
    if ![registry, namespace, name, tag].into_iter().all(safe) {
        return Err(OllamaError::Invalid("installed model path"));
    }
    Ok(root
        .join("manifests")
        .join(registry)
        .join(namespace)
        .join(name)
        .join(tag))
}
pub fn parse_inventory(value: Value) -> Result<Vec<InstalledModel>, OllamaError> {
    #[derive(Deserialize)]
    struct Details {
        quantization_level: Option<String>,
    }
    #[derive(Deserialize)]
    struct Entry {
        name: String,
        digest: String,
        size: u64,
        remote_host: Option<String>,
        remote_model: Option<String>,
        details: Option<Details>,
    }
    #[derive(Deserialize)]
    struct Inventory {
        models: Vec<Entry>,
    }
    let inventory: Inventory =
        serde_json::from_value(value).map_err(|_| OllamaError::Invalid("inventory"))?;
    if inventory.models.len() > 10000 {
        return Err(OllamaError::Invalid("inventory exceeds model limit"));
    }
    let mut result = Vec::new();
    for entry in inventory.models {
        if entry.name.is_empty()
            || entry.name.len() > 256
            || !digest(&entry.digest)
            || entry.size == 0
            || entry.size > 9007199254740991
            || entry
                .details
                .as_ref()
                .and_then(|details| details.quantization_level.as_ref())
                .is_some_and(|value| value.len() > 100)
        {
            return Err(OllamaError::Invalid("inventory entry"));
        }
        if entry
            .remote_host
            .as_ref()
            .is_some_and(|value| !value.is_empty())
            || entry
                .remote_model
                .as_ref()
                .is_some_and(|value| !value.is_empty())
        {
            continue;
        }
        model_path(Path::new(""), &entry.name)?;
        result.push(InstalledModel {
            id: entry.name,
            digest: entry.digest,
            size_bytes: entry.size,
            quant: entry.details.and_then(|details| details.quantization_level),
            context_length: None,
            kv_bytes_per_token: None,
            capabilities: Vec::new(),
        });
    }
    Ok(result)
}
#[derive(Deserialize)]
pub struct Metadata {
    #[serde(default)]
    pub model_info: BTreeMap<String, Value>,
    #[serde(default)]
    pub capabilities: Vec<String>,
    pub parameters: Option<String>,
    pub remote_host: Option<String>,
}
pub fn parse_metadata(value: Value) -> Result<Metadata, OllamaError> {
    let metadata: Metadata =
        serde_json::from_value(value).map_err(|_| OllamaError::Invalid("model metadata"))?;
    if metadata
        .remote_host
        .as_ref()
        .is_some_and(|value| !value.is_empty())
        || metadata.capabilities.len() > 100
        || metadata.capabilities.iter().any(|value| value.len() > 100)
        || metadata
            .parameters
            .as_ref()
            .is_some_and(|value| value.len() > 100000)
    {
        return Err(OllamaError::Invalid("local model metadata"));
    }
    Ok(metadata)
}
pub fn inspect_metadata(
    mut model: InstalledModel,
    value: Value,
) -> Result<InstalledModel, OllamaError> {
    let metadata = parse_metadata(value)?;
    let family = metadata
        .model_info
        .get("general.architecture")
        .and_then(Value::as_str)
        .unwrap_or("");
    let number = |key: &str, maximum: u64| {
        metadata
            .model_info
            .get(&format!("{family}.{key}"))
            .and_then(Value::as_u64)
            .filter(|value| *value > 0 && *value <= maximum)
    };
    model.context_length = number("context_length", 10000000).map(|value| value as u32);
    if ["llama", "qwen2"].contains(&family) {
        model.kv_bytes_per_token = (|| {
            let layers = number("block_count", 100000)?;
            let heads = number("attention.head_count_kv", 100000)?;
            let key = number("attention.key_length", 100000)?;
            let value = number("attention.value_length", 100000)?;
            layers
                .checked_mul(heads)?
                .checked_mul(key + value)?
                .checked_mul(2)
        })();
    }
    model.capabilities = metadata.capabilities;
    Ok(model)
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Layer {
    digest: String,
    size: u64,
    media_type: String,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct Manifest {
    schema_version: u8,
    config: Layer,
    layers: Vec<Layer>,
}
fn contained(root: &Path, path: &Path) -> Result<(), OllamaError> {
    let canonical = path
        .canonicalize()
        .map_err(|_| OllamaError::Integrity("missing local content"))?;
    if !canonical.starts_with(root) {
        return Err(OllamaError::Integrity("local path escaped model root"));
    }
    Ok(())
}
pub async fn verify_manifest(
    root: &Path,
    id: &str,
    expected_manifest: &str,
    expected_weight: Option<&str>,
    expected_bytes: Option<u64>,
    cancel: &CancellationToken,
) -> Result<(), OllamaError> {
    if !digest(expected_manifest)
        || expected_weight.is_some_and(|value| !digest(value))
        || expected_bytes.is_some_and(|value| value == 0 || value > 9007199254740991)
    {
        return Err(OllamaError::Invalid("expected integrity"));
    }
    if cancel.is_cancelled() {
        return Err(OllamaError::Request("cancelled".into()));
    }
    let root = root
        .canonicalize()
        .map_err(|_| OllamaError::Integrity("model root unavailable"))?;
    let path = model_path(&root, id)?;
    contained(&root, &path)?;
    let raw = crate::state::secure_read(&path, 4 * 1024 * 1024, false)
        .map_err(|_| OllamaError::Integrity("manifest unavailable"))?;
    if format!("{:x}", Sha256::digest(raw.as_bytes())) != expected_manifest {
        return Err(OllamaError::Integrity("manifest digest mismatch"));
    }
    let manifest: Manifest =
        serde_json::from_str(&raw).map_err(|_| OllamaError::Invalid("manifest"))?;
    if manifest.schema_version != 2 || manifest.layers.is_empty() || manifest.layers.len() > 10000 {
        return Err(OllamaError::Invalid("manifest layers"));
    }
    let weights: Vec<_> = manifest
        .layers
        .iter()
        .filter(|layer| layer.media_type == "application/vnd.ollama.image.model")
        .collect();
    if weights.is_empty() {
        return Err(OllamaError::Integrity("manifest has no model weights"));
    }
    if expected_weight.is_some_and(|expected| {
        !weights
            .iter()
            .any(|layer| layer.digest == format!("sha256:{expected}"))
    }) {
        return Err(OllamaError::Integrity("catalog digest mismatch"));
    }
    let total = weights
        .iter()
        .try_fold(0u64, |total, layer| total.checked_add(layer.size))
        .ok_or(OllamaError::Invalid("weight size overflow"))?;
    if expected_bytes.is_some_and(|expected| total < expected.div_ceil(2)) {
        return Err(OllamaError::Integrity("catalog size floor"));
    }
    for layer in std::iter::once(&manifest.config).chain(&manifest.layers) {
        let sha = layer
            .digest
            .strip_prefix("sha256:")
            .filter(|value| digest(value))
            .ok_or(OllamaError::Invalid("blob digest"))?;
        if layer.size > 9007199254740991 {
            return Err(OllamaError::Invalid("blob size"));
        }
        let blob = root.join("blobs").join(format!("sha256-{sha}"));
        contained(&root, &blob)?;
        let (actual, bytes) = crate::acquire::hash_file(&blob, cancel)
            .await
            .map_err(|_| OllamaError::Integrity("blob unavailable or cancelled"))?;
        if actual != sha || bytes != layer.size {
            return Err(OllamaError::Integrity("blob digest or size mismatch"));
        }
    }
    if crate::state::secure_read(&path, 4 * 1024 * 1024, false)
        .map_err(|_| OllamaError::Integrity("manifest unavailable"))?
        != raw
    {
        return Err(OllamaError::Integrity(
            "manifest changed during verification",
        ));
    }
    Ok(())
}

#[async_trait::async_trait]
pub trait InstalledVerifier: Send + Sync {
    async fn verify(
        &self,
        id: &str,
        digest: &str,
        cancel: &CancellationToken,
    ) -> Result<(), OllamaError>;
}
pub struct LocalVerifier {
    pub root: PathBuf,
}
#[async_trait::async_trait]
impl InstalledVerifier for LocalVerifier {
    async fn verify(
        &self,
        id: &str,
        digest: &str,
        cancel: &CancellationToken,
    ) -> Result<(), OllamaError> {
        verify_manifest(&self.root, id, digest, None, None, cancel).await
    }
}
pub struct InstalledModels<'runtime> {
    transport: &'runtime dyn crate::http::Transport,
    verifier: &'runtime dyn InstalledVerifier,
}
impl<'runtime> InstalledModels<'runtime> {
    pub fn new(
        transport: &'runtime dyn crate::http::Transport,
        verifier: &'runtime dyn InstalledVerifier,
    ) -> Self {
        Self {
            transport,
            verifier,
        }
    }
    async fn request(
        &self,
        endpoint: &str,
        path: &str,
        body: Option<Value>,
        cancel: &CancellationToken,
    ) -> Result<Value, OllamaError> {
        let request = crate::http::Request::new(endpoint, path, body, None)
            .map_err(|error| OllamaError::Request(error.to_string()))?;
        crate::http::read_json(self.transport, request, cancel, 4 * 1024 * 1024)
            .await
            .map_err(|error| OllamaError::Request(error.to_string()))
    }
    pub async fn list(
        &self,
        endpoint: &str,
        cancel: &CancellationToken,
    ) -> Result<Vec<InstalledModel>, OllamaError> {
        parse_inventory(self.request(endpoint, "/api/tags", None, cancel).await?)
    }
    pub async fn inspect(
        &self,
        endpoint: &str,
        id: &str,
        cancel: &CancellationToken,
    ) -> Result<InstalledModel, OllamaError> {
        model_path(Path::new(""), id)?;
        let selected = exact(self.list(endpoint, cancel).await?, id)?;
        inspect_metadata(
            selected,
            self.request(
                endpoint,
                "/api/show",
                Some(serde_json::json!({"model":id})),
                cancel,
            )
            .await?,
        )
    }
    pub async fn activate(
        &self,
        endpoint: &str,
        model: &InstalledModel,
        context: Option<u32>,
        cancel: &CancellationToken,
    ) -> Result<String, OllamaError> {
        self.activate_with_id(
            endpoint,
            model,
            context,
            &uuid::Uuid::new_v4().to_string(),
            cancel,
        )
        .await
    }
    pub async fn activate_with_id(
        &self,
        endpoint: &str,
        model: &InstalledModel,
        context: Option<u32>,
        unique: &str,
        cancel: &CancellationToken,
    ) -> Result<String, OllamaError> {
        crate::state::loopback(endpoint).map_err(|_| OllamaError::Invalid("endpoint"))?;
        model_path(Path::new(""), &model.id)?;
        if !digest(&model.digest)
            || context.is_some_and(|context| !(1..=10000000).contains(&context))
        {
            return Err(OllamaError::Invalid("activation metadata"));
        }
        let variant = context.map(|context| format!("llmup-context-{unique}:{context}"));
        if let Some(variant) = &variant {
            model_path(Path::new(""), variant)?;
            if variant.len() > 256 {
                return Err(OllamaError::Invalid("variant name"));
            }
        }
        self.verifier
            .verify(&model.id, &model.digest, cancel)
            .await?;
        if exact(self.list(endpoint, cancel).await?, &model.id)?.digest != model.digest {
            return Err(OllamaError::Integrity("source manifest changed; retry"));
        }
        let Some(context) = context else {
            return Ok(model.id.clone());
        };
        let variant = variant.ok_or(OllamaError::Invalid("variant name"))?;
        let created=self.request(endpoint,"/api/create",Some(serde_json::json!({"model":variant,"from":model.id,"parameters":{"num_ctx":context},"stream":false})),cancel).await?;
        if created.get("status").and_then(Value::as_str) != Some("success") {
            return Err(OllamaError::Request(
                "context variant creation failed".into(),
            ));
        }
        let metadata = parse_metadata(
            self.request(
                endpoint,
                "/api/show",
                Some(serde_json::json!({"model":variant})),
                cancel,
            )
            .await?,
        )?;
        if !metadata.parameters.as_ref().is_some_and(|parameters| {
            parameters.lines().any(|line| {
                let mut parts = line.split_whitespace();
                parts.next() == Some("num_ctx")
                    && parts.next().and_then(|value| value.parse::<u32>().ok()) == Some(context)
            })
        }) {
            return Err(OllamaError::Integrity(
                "runtime did not retain requested context",
            ));
        }
        let installed = exact(self.list(endpoint, cancel).await?, &variant)?;
        self.verifier
            .verify(&installed.id, &installed.digest, cancel)
            .await?;
        if exact(self.list(endpoint, cancel).await?, &model.id)?.digest != model.digest {
            return Err(OllamaError::Integrity(
                "source changed during context activation",
            ));
        }
        Ok(variant)
    }
}
fn exact(models: Vec<InstalledModel>, id: &str) -> Result<InstalledModel, OllamaError> {
    let mut matching = models.into_iter().filter(|model| model.id == id);
    let selected = matching
        .next()
        .ok_or(OllamaError::Invalid("model not installed locally"))?;
    if matching.next().is_some() {
        return Err(OllamaError::Invalid("ambiguous installed model"));
    }
    Ok(selected)
}
